File: src/main/java/com/github/wolray/line/io/LineWriter.java
package com.github.wolray.line.io;

import com.alibaba.fastjson.JSON;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.LinkedList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.function.Function;

/**
 * @author wolray
 */
public class LineWriter<T> {
    private final Function<T, String> formatter;

    public LineWriter(Function<T, String> formatter) {
        this.formatter = formatter;
    }

    public static <T> LineWriter<T> byJson() {
        return new LineWriter<>(JSON::toJSONString);
    }

    public static <T> CsvWriter<T> byCsv(String sep, Class<T> type) {
        return new CsvWriter<>(new ValuesJoiner<>(new TypeValues<>(type)), sep);
    }

    public Session write(String file) {
        return new Session(file);
    }

    public class Session {
        protected final String file;
        private final List<String> headers = new LinkedList<>();
        private boolean append;

        protected Session(String file) {
            this.file = file;
        }

        public Session addHeader(String header) {
            headers.add(header);
            return this;
        }

        public Session appendToFile() {
            append = true;
            return this;
        }

        protected void preprocess(BufferedWriter bw) throws IOException {}

        public void asyncWith(Iterable<T> iterable) {
            CompletableFuture.runAsync(() -> with(iterable));
        }

        public void with(Iterable<T> iterable) {
            try (BufferedWriter bw = new BufferedWriter(new FileWriter(file, append))) {
                if (!append) {
                    preprocess(bw);
                    for (String header : headers) {
                        bw.write(header);
                        bw.newLine();
                    }
                }
                for (T t : iterable) {
                    bw.write(formatter.apply(t));
                    bw.newLine();
                }
            } catch (IOException e) {
                throw new UncheckedIOException(e);
            }
        }
    }
}

File: src/main/java/com/github/wolray/line/io/LineWriter.kt
package com.github.wolray.line.io

import com.alibaba.fastjson.JSON
import java.io.BufferedWriter
import java.io.FileWriter
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.function.Function

/**
 * @author wolray
 */
open class LineWriter<T>(private val formatter: Function<T, String>) {

    open fun write(file: String) = Session(file)

    open inner class Session(protected val file: String) : Chainable<Session> {
        private val headers = LinkedList<String>()
        private var append = false
        override val self get() = this

        fun addHeader(header: String) = apply { headers.add(header) }
        fun appendToFile() = apply { append = true }

        open fun markUtf8() = this
        open fun autoHeader() = this
        open fun columnNames(vararg names: String) = this

        protected open fun preprocess(bw: BufferedWriter) {}

        fun asyncWith(iterable: Iterable<T>) {
            CompletableFuture.runAsync { with(iterable) }
        }

        fun with(iterable: Iterable<T>) {
            BufferedWriter(FileWriter(file, append)).use { bw ->
                if (!append) {
                    preprocess(bw)
                    headers.forEach { bw.writeLine(it) }
                }
                iterable.forEach { bw.writeLine(formatter.apply(it)) }
            }
        }
    }

    companion object {
        @JvmStatic
        fun <T> byJson() = LineWriter<T> { JSON.toJSONString(it) }

        @JvmStatic
        fun <T> byCsv(sep: String, type: Class<T>): CsvWriter<T> {
            return CsvWriter(ValuesJoiner(TypeValues(type)), sep)
        }

        private fun BufferedWriter.writeLine(s: String) {
            write(s)
            newLine()
        }
    }
}
